src/github.com/matrix-org/dendrite/clientapi/routing/room_tagging.go

```diff
@@ -0,0 +1,186 @@
+		return util.JSONResponse{
+			Code: http.StatusMethodNotAllowed,
+			JSON: jsonerror.NotFound("Bad method"),
+		}
+	}
+
+	mtag := NewMTag()
+
+	localpart, _, err := gomatrixserverlib.SplitID('@', userID)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	data, err := accountDB.GetAccountDataByType(
+		req.Context(), localpart, roomID, "m.tag",
+	)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	dataByte, _ := json.Marshal(data)
+	err = json.Unmarshal(dataByte, &mtag)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: mtag,
+	}
+}
+
+// PutTag implements PUT /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags/{tag}
+func PutTag(req *http.Request, accountDB *accounts.Database, userID string, roomID string, tag string) util.JSONResponse {
+
+	if req.Method != http.MethodPut {
+		return util.JSONResponse{
+			Code: http.StatusMethodNotAllowed,
+			JSON: jsonerror.NotFound("Bad method"),
+		}
+	}
+
+	localpart, _, err := gomatrixserverlib.SplitID('@', userID)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	//Check for existing entries of tags for this ROOM ID and localpart
+	data, err := accountDB.GetAccountDataByType(
+		req.Context(), localpart, roomID, "m.tag",
+	)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	mtag := NewMTag()
+
+	var properties common.TagProperties
+
+	if reqErr := httputil.UnmarshalJSONRequest(req, &properties); reqErr != nil {
+		return *reqErr
+	}
+
+	if len(data) > 0 {
+		dataByte, _ := json.Marshal(data)
+		err = json.Unmarshal(dataByte, &mtag)
+		if err != nil {
+			return httputil.LogThenError(req, err)
+		}
+	}
+
+	mtag.Tags[tag] = properties
+
+	newTagData, _ := json.Marshal(mtag)
+
+	if err := accountDB.SaveAccountData(
+		req.Context(), localpart, roomID, "m.tag", string(newTagData),
+	); err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: struct{}{},
+	}
+}
+
+// DeleteTag implements DELETE /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags/{tag}
+func DeleteTag(req *http.Request, accountDB *accounts.Database, userID string, roomID string, tag string) util.JSONResponse {
+
+	if req.Method != http.MethodDelete {
+		return util.JSONResponse{
+			Code: http.StatusMethodNotAllowed,
+			JSON: jsonerror.NotFound("Bad method"),
+		}
+	}
+
+	localpart, _, err := gomatrixserverlib.SplitID('@', userID)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	data, err := accountDB.GetAccountDataByType(
+		req.Context(), localpart, roomID, "m.tag",
+	)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	mtag := NewMTag()
+
+	if len(data) > 0 {
+		dataByte, _ := json.Marshal(data)
+		err = json.Unmarshal(dataByte, &mtag)
+		if err != nil {
+			return httputil.LogThenError(req, err)
+		}
+	} else {
+		//Error indicating there is no Tag data
+		return util.JSONResponse{}
```

Comment:
For an error, you should return a specific error.
Empty response is good only when we have a success.
Same goes for subsequent lines.

Reply:
Sure, I'll be changing that.